### ISSUE TITLE
Outbound http requests leak unmanaged memory on Alpine 3.9

### ISSUE THREAD

**Issue by Bio2hazard** [dotnet/corefx#41875]
Copied from https://github.com/aspnet/AspNetCore/issues/14117

TL;DR:
Using HttpClient to make outbound calls in any of the alpine3.9 docker images for .net core 2.2.7 or 3.0 results in a steady, unbounded growth of working set / unmanaged memory.

Alpine 3.8 and Alpine 3.10 Previews ( as reported by @GREsau in the original issue ) are not affected.

Please see the original issue for charts and more context.

Here is a solution that exhibits the the growing working set size: 
[Alpine39MemoryRepro.zip](https://github.com/aspnet/AspNetCore/files/3736935/Alpine39MemoryRepro.zip)

After 4 hours of runtime:
Alpine 3.8's working set size remains between 98-101MB
Alpine 3.9's working set size started at ~98 MB, and is now at 107 MB. 
GC Generations 0, 1 and 2 are all being collected and the managed memory has remained stable.

After 24 hours of runtime:
Alpine 3.8 is still at 98-101MB
Alpine 3.9 is now at almost 150 MB.

Usage for the repro:
1. Navigate to the root of the extracted zip file
1. Build 3.9 Image via: `docker build -t alpinememory -f .\Dockerfile-localtest .`
1. (Optional) Build 3.8 Image via: `docker build -t alpinememory38 -f .\Dockerfile-localtest38 .`
1. Run 3.9 Image via: `docker run -p 5001:8080 --name alpinememory -d alpinememory` ( alter host port if you wish )
1. (Optional) Run 3.8 Image via: `docker run -p 5002:8080 --name alpinememory38 -d alpinememory38`
1. Wait at least 3 hours for trends to emerge
1. Grab a report of working set size via `docker cp alpinememory:/tmp/report.csv ./report_39.csv`
1. (Optional) Grab a report of alpine 3.8 for comparison via `docker cp alpinememory38:/tmp/report.csv ./report_38.csv`

What the repro solution does:
- Relatively standard aspnet core 2.27 web api style project, with nginx reverse proxy.
- Starts 2 IHostedServices that inherit from `BackgroundWorker`.
- `Worker/ReportingWorker.cs` runs every minute and dumps memory-related information from `Process.` and `GC.` namespaces to a text file, and only exists to conveniently observe memory behavior over time without requiring any additional dependencies
- `Worker/WebRequestWorker.cs` runs every second and sends a HTTPS Get to one out of 16 pre-configured URLs. It then logs the number of A or a chars in the response, pretty sure that's unnecessary but I wanted to make sure the response content was used in some manner and nothing was optimized out.
   - The URLs were chosen from the Alexa page ranks since I wanted to make sure to only send GETs to big, established sites. 1 request every 16 seconds probably does not bother them. 👍 
   - The requests are made using the recommended pattern of using `services.AddHttpClient();` in `Startup.cs`, injecting `IHttpClientFactory` and using `CreateClient` on-demand.

Here are my results after 4 hours:
Alpine 3.9: [report_39.txt](https://github.com/aspnet/AspNetCore/files/3736982/report_39.txt)
Alpine 3.8: [report_38.txt](https://github.com/aspnet/AspNetCore/files/3736983/report_38.txt)

Results after close to 24 hours:
Alpine 3.9: [report_39.txt](https://github.com/dotnet/corefx/files/3741317/report_39.txt)
Alpine 3.8: [report_38.txt](https://github.com/dotnet/corefx/files/3741318/report_38.txt)

We are subject to time limits when it comes to applying security patches, and the OS of the container instance is no exception. As 3.8 is quite old, this issue blocks us from updating and may force us to switch to the bionic images instead, which would be pretty disappointing as we are otherwise very happy with Alpine. Thank you for your time! 👍

**Comment by davidsh** [comment]
Looking at your code in WebRequestWorker.cs, I see this:

```c#
                    if (index >= _uris.Length)
                    {
                        index = 0;
                    }

                    var uri = _uris[index];
                    index++;


                    var client = _httpClientFactory.CreateClient("RequestWorker");
                    var cts = new CancellationTokenSource(2_000);
                    var pageResponse = await client.GetAsync(uri, cts.Token);

                    pageResponse.EnsureSuccessStatusCode();

                    var pageContent = await pageResponse.Content.ReadAsStringAsync();

                    var countOfA = 0;
                    var pos = 0;
                    while ((pos = pageContent.IndexOfAny(new []{'a', 'A'}, pos + 1)) != -1) countOfA++;


                    _logger.LogInformation($"{uri}'s response contains {countOfA} A/a");
```

Patterns like this:

```c#
var pageResponse = await client.GetAsync(uri, cts.Token);
```

will cause memory to climb and TCP connections to hang around a long time.  The code pattern should be more like this:

```c#
using (HttpResponseMessage pageResponse = await client.GetAsync(uri, cts.Token))
{
    // Do work with the HttpResponseMessage and HttpResponseMessage.Content etc.
}
```

**Comment by Bio2hazard** [comment]
The production code where this issue first surfaced properly wraps everything in using blocks. I apologize for omitting them in the reproduction code. 

I am fairly confident that memory will leak even with proper disposal, and it also does not explain the difference in behavior between alpine 3.8 and alpine 3.9. Just in case however I will apply proper using blocks in my repro code and run it to verify. I will report back with my findings.

**Comment by davidsh** [comment]
>Alpine 3.8 and Alpine 3.10 Previews ( as reported by @GREsau in the original issue ) are not affected.

So, this also tells me that perhaps there is a GC regression of some kind in Alpine 3.9 image.

**Comment by davidsh** [comment]
@wfurt @stephentoub

**Comment by wfurt** [comment]
Since the the claim is that 3.8 and 3.10 works fine and we have same binaries this may be OS bug. We would need to figure out what consumes that used memory to know it for sure.

Since you have repro @Bio2hazard , it would be interesting to know if this also reproduces with any Tls e.g. use http:// instead of https:// in your site list.   

cc: @janvorli

**Comment by Bio2hazard** [comment]
Still running the repro with proper disposal pattern, will report back soon.

@wfurt I'll try to run the repro with http, but once problem is that almost all sites automatically redirect to https these days ( which is good! but bad for this experiment ).

I need to figure out how to provide a custom handler to the named http client registration in startup.cs, so I can disable redirects, or find a good set of http uri's to test with.

I will report back once I have results re: http vs https

**Comment by wfurt** [comment]
Disable auto-redirect https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclienthandler.allowautoredirect?view=netcore-3.0
Even if that fails, or you use smaller URL set it may be ok. 

I think for purpose of this that should be ok as we are trying to check if this is some GC bug or strange OS dependency. 

BWT I assume you use default handler e.g curl is out of the picture. 

You can also try fetching the list on loop to split server vs client side operation.   
I think asp.net now also default to managed implementation so there should be very little platform code.

**Comment by thomasf-nvm** [comment]
I work with gnjack and GREsau. We’ve also been working on a separate repro,
and while I don’t yet have conclusive evidence that HTTPS leaks and HTTP
doesn’t, we have seen behaviour that supports that theory.

We will try to independently confirm that over the next few days.

**Comment by Bio2hazard** [comment]
@davidsh I had it running with proper disposal on the http response message and the cancellation token source for 2 hours now and while the overall memory usage was reduced by about 2 MB for both 3.8 and 3.9, 3.9 is still leaking at the same rate as before.

@wfurt Oh I know how to disable redirects on the handler. It's more that the handler can only be configured AFAIK when you create a new HttpClient, but with `services.AddHttpClient();` you do not have that option. It's probably possible with a typed http client, but instead I'll just fine some http capable targets and not worry about the redirects 👍 

Yes, the default handler is used. 

Regarding fetching the list on loop; I'm not sure I fully understand. 

The minimal repro has a IHostedService that loops through all uris on the list, so it's already fetching every entry in the list on a loop, but I'm not sure how that helps split server vs. client side operation.

I will report back after I have a good set of data using non-redirecting http uris.

**Comment by Bio2hazard** [comment]
@wfurt Outbound requests not using Tls do not seem to leak memory, so it definitely seems isolated to https requests only.

[report_http_only_38.txt](https://github.com/dotnet/corefx/files/3745516/report_http_only_38.txt)
[report_http_only_39.txt](https://github.com/dotnet/corefx/files/3745517/report_http_only_39.txt)

**Comment by wfurt** [comment]
good to know. It was just guess but we now have more specific focus. 
With the "fetch in loop" the idea is to go back to old good console app and remove asp.net and  nginx from the picture. If the leak is in ssl that should also reproduce but it should be much easier to debug.

**Comment by Bio2hazard** [comment]
Hey @wfurt I made a simple reproduction using a console app and a loop in the main function as you suggested.

I tried .net core 3.0 but could not reproduce the issue there. Not with `mcr.microsoft.com/dotnet/core/runtime:3.0-alpine` or with `mcr.microsoft.com/dotnet/core/runtime:3.0-alpine3.9`

However, once I switched to .net core 2.2.7 on alpine3.9 the issue immediately reappeared ( using  `mcr.microsoft.com/dotnet/core/runtime:2.2.7-alpine3.9` ).

Repro code can be found at [this gist](https://gist.github.com/Bio2hazard/4e1dcf897a15dcbc0d79533c2f6692f9)

3.8 staying steady at 48 MB Working Set Size: [report_reuse_38.txt](https://github.com/dotnet/corefx/files/3747447/report_reuse_38.txt)
3.9 increasing steadily to 58 MB Working Set Size after 8 hours: [report_reuse_39.txt](https://github.com/dotnet/corefx/files/3747448/report_reuse_39.txt)

**Comment by karelz** [comment]
Looks like it is addressed in 3.0+. Given that 2.2 is not supported anymore, we can close it.
